music21/stream/tools.py

```diff
@@ -0,0 +1,170 @@
+        {0.0} <music21.note.Note C>
+        {1.0} <music21.note.Note C>
+        {2.0} <music21.note.Note D>
+    {3.0} <music21.stream.Measure 2 offset=3.0>
+        {0.0} <music21.clef.TrebleClef>
+        {0.0} <music21.key.KeySignature of 6 sharps>
+        {0.0} <music21.meter.TimeSignature 3/4>
+        {0.0} <music21.note.Note E>
+        {1.0} <music21.note.Note F>
+        {2.0} <music21.note.Note G>
+    {6.0} <music21.stream.Measure 3 offset=6.0>
+        {0.0} <music21.clef.BassClef>
+        {0.0} <music21.key.KeySignature of 5 flats>
+        {0.0} <music21.meter.TimeSignature 2/4>
+        {0.0} <music21.note.Note A>
+        {1.0} <music21.note.Note B>
+    {8.0} <music21.stream.Measure 4 offset=8.0>
+        {0.0} <music21.note.Note C>
+        {1.0} <music21.bar.Barline type=final>
+
+    Calling removeDuplicates should remove the duplicates
+    even with those changes now stored within measures,
+    not directly on the part.
+    Specifically, in our example,
+    the duplicates entries are removed from measure 2
+    and the actual changes in measure 3 remain.
+
+    >>> s = stream.tools.removeDuplicates(s)
+    >>> s.show('t')
+    {0.0} <music21.stream.Measure 1 offset=0.0>
+        {0.0} <music21.clef.TrebleClef>
+        {0.0} <music21.key.KeySignature of 6 sharps>
+        {0.0} <music21.meter.TimeSignature 3/4>
+        {0.0} <music21.note.Note C>
+        {1.0} <music21.note.Note C>
+        {2.0} <music21.note.Note D>
+    {3.0} <music21.stream.Measure 2 offset=3.0>
+        {0.0} <music21.note.Note E>
+        {1.0} <music21.note.Note F>
+        {2.0} <music21.note.Note G>
+    {6.0} <music21.stream.Measure 3 offset=6.0>
+        {0.0} <music21.clef.BassClef>
+        {0.0} <music21.key.KeySignature of 5 flats>
+        {0.0} <music21.meter.TimeSignature 2/4>
+        {0.0} <music21.note.Note A>
+        {1.0} <music21.note.Note B>
+    {8.0} <music21.stream.Measure 4 offset=8.0>
+        {0.0} <music21.note.Note C>
+        {1.0} <music21.bar.Barline type=final>
+
+    '''
+
+    supportedClasses = [meter.TimeSignature, key.KeySignature, clef.Clef]
+    if classesToRemove is None:
+        classesToRemove = supportedClasses
+
+    listOfObjectsToRemove = []
+
+    for thisClass in classesToRemove:
+
+        if thisClass not in supportedClasses:
+            raise ValueError(f'Invalid class. Only {supportedClasses} are supported.')
+
+        allStates = thisStream.recurse().getElementsByClass(thisClass)
+
+        if len(allStates) < 2:  # Not used, or doesn't change
+            continue
+
+        currentState = allStates[0]  # First to initialize: can't be a duplicate
+        for thisState in allStates[1:]:
+            if thisState == currentState:
+                listOfObjectsToRemove.append(thisState)
+            else:
+                currentState = thisState
+
+    for item in listOfObjectsToRemove:
+        m = item.getContextByClass(stream.Measure)
+        m.remove(item, recurse=True)
```

Comment:
It's not safe to assume the item exists in a measure. You might try building a dict keyed on streams (activeSites of the objects encountered), and then doing one call to remove() per activeSite, providing a list of elements.

Reply:
https://github.com/cuthbertLab/music21/pull/1454/commits/7c85ad3866826ecf817b0efb5604e648aae04a16